File: app/src/main/java/io/github/trojan_gfw/igniter/qrcode/QRCodeDecoder.java
package io.github.trojan_gfw.igniter.qrcode;

import android.graphics.Bitmap;
import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.google.mlkit.vision.barcode.common.Barcode;
import com.google.mlkit.vision.barcode.BarcodeScanner;
import com.google.mlkit.vision.barcode.BarcodeScannerOptions;
import com.google.mlkit.vision.barcode.BarcodeScanning;
import com.google.mlkit.vision.common.InputImage;

import java.util.Optional;

public class QRCodeDecoder {
    private final BarcodeScanner mScanner;

    public QRCodeDecoder() {
        BarcodeScannerOptions options = new BarcodeScannerOptions.Builder()
                .setBarcodeFormats(Barcode.FORMAT_QR_CODE).build();
        mScanner = BarcodeScanning.getClient(options);
    }

    public void decode(@NonNull Bitmap bitmap, @NonNull OnQRCodeDecodeListener listener) {
        mScanner.process(InputImage.fromBitmap(bitmap, 0)).addOnSuccessListener(barcodes -> {
            Optional<Barcode> qrCode = barcodes.stream().filter(barcode -> {
                String rawValue = barcode.getRawValue();
                return rawValue != null && !TextUtils.isEmpty(rawValue.trim());
            }).findFirst();
            if (qrCode.isPresent()) {
                listener.onSuccess(qrCode.get().getRawValue());
            } else {
                listener.onNoQRCode();
            }
        }).addOnCanceledListener(listener::onNoQRCode)
                .addOnFailureListener(listener::onError)
                .addOnCompleteListener(barcodes -> listener.onComplete());
    }

    public void close() {
        mScanner.close();
    }

    public interface OnQRCodeDecodeListener {
        void onSuccess(String qrCode);

        void onError(Exception e);

        void onNoQRCode();

        void onComplete();
    }
}

File: app/src/main/java/io/github/trojan_gfw/igniter/qrcode/QRCodeDecoder.kt
package io.github.trojan_gfw.igniter.qrcode

import android.graphics.Bitmap
import android.text.TextUtils
import androidx.annotation.NonNull
import com.google.mlkit.vision.barcode.Barcode
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage

class QRCodeDecoder {
    private val scanner = BarcodeScanning.getClient(BarcodeScannerOptions.Builder()
            .apply { setBarcodeFormats(Barcode.FORMAT_QR_CODE) }.build())

    fun decode(bitmap: Bitmap, listener: OnQRCodeDecodeListener) {
        scanner.process(InputImage.fromBitmap(bitmap, 0)).addOnSuccessListener { barcodes ->
            val barcode = barcodes.firstOrNull {
                it.rawValue != null && !TextUtils.isEmpty(it.rawValue!!.trim())
            }
            if (null == barcode) {
                listener.onNoQRCodeDetected()
            } else {
                listener.onSuccess(barcode.rawValue!!)
            }
        }.addOnFailureListener { exception ->
            listener.onError(exception)
        }.addOnCompleteListener {
            listener.onComplete()
        }.addOnCanceledListener { listener.onNoQRCodeDetected() }
    }

    fun close() {
        scanner.close()
    }

    interface OnQRCodeDecodeListener {
        fun onSuccess(@NonNull result: String)
        fun onNoQRCodeDetected()
        fun onError(@NonNull exception: Exception)
        fun onComplete()
    }
}